src/editor/validator.ts

```diff
@@ -361,7 +408,61 @@ export class Validator {
         return context.selected && context.token.isEmpty && context.token instanceof TypedEmptyExpr;
     }
 
-    private getStatementAbove(providedContext?: Context): Statement {
+    private getPrevSibling(providedContext?: Context): Statement {
+        const context = providedContext ? providedContext : this.module.focus.getContext();
+
+        return this.getStatementInBody(
+            context.lineStatement.rootNode as Statement | Module,
+            context.lineStatement.indexInRoot - 1
+        );
+    }
+
+    private getNextSibling(providedContext?: Context): Statement {
+        const context = providedContext ? providedContext : this.module.focus.getContext();
+
+        return this.getStatementInBody(
+            context.lineStatement.rootNode as Statement | Module,
+            context.lineStatement.indexInRoot + 1
+        );
+    }
+
+    private getNextSiblingOfRoot(providedContext?: Context): Statement {
+        const context = providedContext ? providedContext : this.module.focus.getContext();
+        const curRoot = context.lineStatement.rootNode;
+
+        if (curRoot instanceof Statement) {
+            return this.getStatementInBody(curRoot.rootNode as Statement | Module, curRoot.indexInRoot + 1);
+        }
+
+        return null;
+    }
+
+    private getPrevSiblingOfRoot(providedContext?: Context): Statement {
+        const context = providedContext ? providedContext : this.module.focus.getContext();
+        const curRoot = context.lineStatement.rootNode;
+
+        if (curRoot instanceof Statement) {
+            return this.getStatementInBody(curRoot.rootNode as Statement | Module, curRoot.indexInRoot - 1);
+        }
+
+        return null;
+    }
+
+    private getStatementInBody(bodyContainer: Statement | Module, index: number): Statement {
```

Comment:
Is this not the same as the value of indexInRoot for a given statement?

Reply:
it is, but this particular method is index-out-of-bounds-safe.

Reply:
I see.